### DISCUSSION TITLE
Enable/Disable MediaElement foreground service.

### DISCUSSION THREAD

**Discussion by Pmr-precure** [CommunityToolkit/Maui#2225] [New Feature Discussions]
I would like to request an option to disable the foreground service when using MediaElement on Android. Currently, MediaElement initiates a foreground service by default during media playback.
In our case we use MediaElement for tutorials and already have a Foreground service running with information about our connected BLE devices. In that case it would be nice not having the MediaElements notification aswell.

**Comment by mrucker** [comment]
I agree, this wouldn't even be incredibly difficult. For example, the `ServiceAttribute` could be removed from `MediaControlsService`, and instructions could be added to the documentation instructing users to add the appropriate service and intent filters to the manifest if they want it. Also, you would need to remove the foreground permission declarations at the top of `MauiMediaElement`.

**Reply by ne0rrmatrix** [reply]
If we use init and add it as builder option we can just choose to never call the service and it will never use the `ServiceAttribute`. We have a MediaElementOption class now that we can do this in.

**Comment by ne0rrmatrix** [comment]
We discussed this today in our monthly meeting. I will be opening a discussion to implement this feature as it wanted by the community and should be easy to implement. If you watch the youtube stream you will see it at the end of the stream near the very end.

**Reply by mrucker** [reply]
That's great! Yes, in our app we only have one video in a single short introduction flow. So, requesting full foreground media playback permissions from the Play store is a little overkill for us. This is especially true for us since Google requires video documentation justifying why foreground permissions are required. That's actually the real rub, if Google didn't have onerous documentation requirements for foreground services we'd probably just roll with it and leave the permissions in place. Because of this we've been creating our own custom MediaElement nuget package from your releases with the two changes I note above. It only takes us like 5 or 10 minutes to make the change.

**Reply by mrucker** [reply]
Oh also, I'm happy to provide a PR with this modification, if your development team is open to it. I'd just ask for a little direction on how you'd like to make this behavior optional. Since AppManifest is made at compile time it seems like a compiler directive is the only way to give developers the option to turn this on/off (assuming the team still wants it on by default).

**Reply by ne0rrmatrix** [reply]
You can add the manifest changes if you intend to use it. If not you can choose not to and then it will just not use it. Using opt in means you only need to set manifest if u intend to use it. If you do it that way you can just set the service to start or stop on demand and this will enable and disable both the service and notifications. You still need to decide at compile time if you do not want to use it all and never request the permissions. If you intend to use it only some of the time you will need to ask for the permissions and add the service in manifest. You can then choose to use the service by opting in using the binding.

This is how I view using it based on feedback I have received. Do you like this idea?

**Reply by mrucker** [reply]
Yes that seems reasonable to me. My only thought is for the CommunityToolkit team. I assume the more common use case is wanting the permissions turned on. So, what you've proposed means that the default setting is probably going to be the more rare use case. If that's the case I'm afraid you guys might get flooded with people complaining about the MediaElement not working, when the problem is simply that users didn't read all the documentation. This is especially true if people upgrade to newer versions and the default permissions have changed.

**Reply by bijington** [reply]
I appreciate your thoughts on reducing the load for the team. I personally think it has to be opt-in unless we can somehow find a way to dynamically load in the permissions for foreground service at compile time. We could add plenty more documentation to make it clear what is initialised by default and how to swap to using the foreground service.

**Reply by mrucker** [reply]
Yes, the only other solution I can imagine to maintain the current default is to wrap the permission attributes with a compiler directive in MauiMediaElement. Then developers who want to turn them off could be instructed to add a compiler directive to their csproj file, Everyone else would get the same default behavior that they get now. I'm not crazy about using compiler directives, but they are kind of a necessary evil in a lot of MAUI development. The directives could look something like the below:

```
<PropertyGroup Condition="$(TargetFramework.Contains('-android'))">
     <DefineConstants>$(DefineConstants);NO_FOREGROUND_PLAYBACK</DefineConstants>
</PropertyGroup>
```

**Reply by bijington** [reply]
I like the idea but it sadly won't work because the compiler directives apply to the code being compiled - for app developers they are only compiling their code, the toolkit will already have been compiled so the directive would be too late.

I like the idea of possibly sticking the permission attributes in our codebase which will only be retained if the developer refers to the code that has the attributes applied. I don't know if this is possible though because these types of attributes are applied to the assembly I think. 

It would be amazing for developers to write something like:

```csharp
builder.UseCommunityToolkitMedia(
    options => options.EnableForegroundService = true);
```

Perhaps some form of source generation could help to solve this

**Reply by mrucker** [reply]
Ha ha. Of course 😄. My brain went for an obvious solution and skipped right past the fact that Nuget packages are deployed as compiled code. I agree, configuring the foreground service permission through the existing builder API would be great.

I've been doing a lot of reading. I think it might be possible to achieve the desired builder interface if you create a custom BuildTask, but this is far outside of my development experience and expertise. The code that auto-magically turns UsesPermissionAttribute into permissions in the manifest is [here](https://github.com/dotnet/android/blob/b9f6a9d6fc95c6d0128b47d59cf7d15bf14a40a8/src/Xamarin.Android.Build.Tasks/Utilities/ManifestDocument.cs#L901-L910) which is called from this [Build Task](https://github.com/dotnet/android/blob/main/src/Xamarin.Android.Build.Tasks/Tasks/GenerateMainAndroidManifest.cs#L98). Maintaining such a complex solution, however, is probably worse than dealing with the disruption to the community by changing default behavior.

**Reply by bijington** [reply]
I think I would agree that it might be less effort to point developers to the documentation. That said this is a good discussion, perhaps someone else will come in with an alternative suggestion?

A further thought I had was (and this might be more effort still) to introduce a second Media based NuGet package that explicitly adds the foreground service support on top of the existing implementation, this would provide an easy opt-in for developers and hopefully an easy way to automatically register the permissions required.

**Reply by JoacimWall** [reply]
Hi 
Is there any status update on this one? 
//Joacim

**Reply by mrucker** [reply]
Hi @JoacimWall, I'm not sure but, I'll reiterate what I said above in case it helps. It is very easy to remove the permissions in the `CommunityToolkit` code yourself and build a custom package from source. That's what my project has done. It took us like half an hour to get it all hooked up, and we've been good to go for several months now.

**Reply by JoacimWall** [reply]
Hi @ne0rrmatrix , Do you have any plans to revert this so we can opt in/out off foreground Service. 
I really appreciate all the time you put into building this library. But forcing users to use foreground service is not good way for a lib. We show video in our app not play audio in the background. If you want to play only audio there is Plugin.Maui.Audio for this.
@mrucker Yes we can build custom package but I don't think this is the way forward. 

//Thanks Joacim

**Reply by ne0rrmatrix** [reply]
@JoacimWall I have started a Feature proposal here: https://github.com/CommunityToolkit/Maui/issues/2657 If you can take a look and if you have any suggestions or alternatives I would appreciate your input.

**Reply by ne0rrmatrix** [reply]
> Hi @ne0rrmatrix , Do you have any plans to revert this so we can opt in/out off foreground Service. I really appreciate all the time you put into building this library. But forcing users to use foreground service is not good way for a lib. We show video in our app not play audio in the background. If you want to play only audio there is Plugin.Maui.Audio for this. @mrucker Yes we can build custom package but I don't think this is the way forward.
> 
> //Thanks Joacim

There are zero plans to remove the service at this time. I just opened a new proposal to allow users to opt out in builder using Media Element Option class and if they opt out it will never load the service. This would be usable in either builder, xaml, or code behind as I envision it. I have tested it and it works. I have to thank @TheCodeTraveler for his assistance on a previous PR that makes this very easy for me to implement.

**Comment by JoacimWall** [comment]
Hi 
I follow this issue and should this not now been solved?
[MediaElement foreground service](https://github.com/CommunityToolkit/Maui/pull/2658#event-22369745063)

We are using 
CommunityToolkit.Maui.MediaElement Version="8.0.0" 
and 
.UseMauiCommunityToolkitMediaElement(false) 

but we get this error when we try to use MediaElement

Java.Lang.RuntimeException: Unable to create service communityToolkit.maui.media.services: java.lang.SecurityException: Starting FGS with type mediaPlayback callerApp=ProcessRecord{e5b087 28696:com.evry.liberoclub.test/u0a195} targetSDK=36 requires permissions: all of the permissions allOf=true [android.permission.FOREGROUND_SERVICE_MEDIA_PLAYBACK] ---> Java.Lang.SecurityException: Starting FGS with type mediaPlayback callerApp=ProcessRecord{e5b087 28696:com.evry.liberoclub.test/u0a195} targetSDK=36 requires permissions: all of the permissions allOf=true [android.permission.FOREGROUND_SERVICE_MEDIA_PLAYBACK] ---> Android.OS.RemoteException: Remote stack trace: at com.android.server.am.ActiveServices.validateForegroundServiceType(ActiveServices.java:2611) at com.android.server.am.ActiveServices.setServiceForegroundInnerLocked(ActiveServices.java:2322) at com.android.server.am.ActiveServices.setServiceForegroundLocked(ActiveServices.java:1679) at com.android.server.am.ActivityManagerService.setServiceForeground(ActivityManagerService.java:13281) at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:3385) --- End of managed Android.OS.RemoteException stack trace --


The only solution for us is to remove the merge of manifest.
by adding this to the AndroidManifest.xml
xmlns:tools="http://schemas.android.com/tools"
...........
.........
 
        <service android:name="communityToolkit.maui.media.services"  tools:node="remove" />
 </application>

**Reply by ne0rrmatrix** [reply]
Use this:
```
.UseMauiCommunityToolkitMediaElement(isAndroidForegroundServiceEnabled: true, static options =>
{
				options.SetDefaultAndroidViewType(AndroidViewType.TextureView);
})
``` 
in your MauiProgram.cs

**Reply by ne0rrmatrix** [reply]
You can read more about that here in the documentation which was updated a few days after the release. https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/views/mediaelement?tabs=android

**Reply by ne0rrmatrix** [reply]
Please open a bug report. I just confirmed the behaviour in a new app with just media element added!

**Reply by JoacimWall** [reply]
Hi 
We don't want have it on.  

So you think this is a bug and we should create a bug report?
What I can se in the docs related to this is this.

"We do not recommend using TextureView unless you have a specific need for it. It has possible performance related issues if enabled and is recommended only for those that need transparencies and other advanced features."  

So what's you recommendation using your solution or remove the merge of manifest?

//Thanks Joacim

**Reply by ne0rrmatrix** [reply]
This is a workaround for now: 
```
.UseMauiCommunityToolkitMediaElement(isAndroidForegroundServiceEnabled: false, static options =>
            {
                options.SetDefaultAndroidViewType(AndroidViewType.TextureView);
                options.SetIsAndroidForegroundServiceEnabled(false);
            });
```
I tested in our sample app and setting it twice. Once as a parameter and then again in options directly will let you turn it off. It is on by default for some odd reason atm. Setting it twice is the workaround.

**Reply by ne0rrmatrix** [reply]
Oh you can just switch off texture view or just omit it. That was me just copy pasting stuff. It will have no impact on service initialization. My bad. You can just use this: 
```
.UseMauiCommunityToolkitMediaElement(isAndroidForegroundServiceEnabled: false, static options =>
            {
                options.SetIsAndroidForegroundServiceEnabled(false);
            });
```
Sorry about the confusion!

**Reply by JoacimWall** [reply]
Ok 
Thanks for your quick response now it works.

//Joacim

**Comment by ne0rrmatrix** [comment]
I'm closing this as completed by #2658
